File: src/main/java/me/meta4245/betterthanmodern/mixin/SheepEntityMixin.java
package me.meta4245.betterthanmodern.mixin;

import me.meta4245.betterthanmodern.event.ItemRegistry;
import me.meta4245.betterthanmodern.mixin.accessor.EntityAccessor;
import net.minecraft.block.Block;
import net.minecraft.entity.passive.SheepEntity;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

import java.util.Random;

@Mixin(SheepEntity.class)
public abstract class SheepEntityMixin {
    @Inject(
            method = "getDroppedItemId",
            at = @At("HEAD"),
            cancellable = true
    )
    private void getDroppedItemId(CallbackInfoReturnable<Integer> cir) {
        EntityAccessor accessor = (EntityAccessor) this;
        Random random = accessor.getRandom();

        int muttonAmount = random.nextInt(1, 3);
        int id;

        if (accessor.getFireTicks() > 0) {
            id = ItemRegistry.cookedMutton.id;
        } else {
            id = ItemRegistry.rawMutton.id;
        }

        accessor.callDropItem(Block.WOOL.id, 1);
        accessor.callDropItem(id, muttonAmount);

        // LivingEntity doesn't drop if getDroppedId returns 0
        cir.setReturnValue(0);
    }
}

File: src/main/kotlin/me/meta4245/betterthanmodern/mixin/SheepEntityMixin.kt
package me.meta4245.betterthanmodern.mixin

import me.meta4245.betterthanmodern.event.ItemRegistry
import me.meta4245.betterthanmodern.mixin.accessor.EntityAccessor
import net.minecraft.entity.passive.SheepEntity
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable

@Mixin(SheepEntity::class)
abstract class SheepEntityMixin {
    @Inject(at = [At("HEAD")], method = ["getDroppedId"], cancellable = true)
    fun getDroppedId(cir: CallbackInfoReturnable<Int?>) {
        val accessor = this as EntityAccessor
        val fireTicks = accessor.fireTicks

        val rawMuttonId: Int
        val cookedMuttonId: Int

        try {
            rawMuttonId =
                ItemRegistry::class.java.getDeclaredField("RawMuttonId")[this] as Int
            cookedMuttonId =
                ItemRegistry::class.java.getDeclaredField("CookedMuttonId")[this] as Int
        } catch (e: Exception) {
            throw RuntimeException(e)
        }

        cir.returnValue = if (fireTicks > 0)
            cookedMuttonId
        else
            rawMuttonId
    }
}
